File: src/commands/vs.rs
use std::sync::Arc;

use anyhow::Result;
use serde_json::{json, Value};

use crate::commands::build::BuildMode;
use crate::workspace;

type Tree = Arc<prodash::tree::Root>;

pub async fn cmd_vs(tree: &Tree) -> Result<()> {
    let mut p = tree.add_child("generate VS/VSCode configs");
    let targets = workspace::get_targets()?;
    let ws = workspace::root_dir();

    let vs_dir = ws.join(".vs");
    let vscode_dir = ws.join(".vscode");
    std::fs::create_dir_all(&vs_dir)?;
    std::fs::create_dir_all(&vscode_dir)?;

    let mut launch_vs = json!({ "version": "0.2.1", "defaults": {}, "configurations": [] });
    let mut tasks_vs = json!({ "version": "0.2.1", "tasks": [] });
    let mut launch_vsc = json!({ "version": "0.2.0", "configurations": [] });
    let mut tasks_vsc = json!({ "version": "2.0.0", "tasks": [] });

    for target in &targets {
        for mode in [BuildMode::Debug, BuildMode::Release] {
            let name = format!("{} ({mode})", target.name);
            let out_rel = format!(
                "build/{}/{}_{}",
                target.path.to_string_lossy(),
                target.name,
                mode
            );

            let out_win = std::path::Path::new(&out_rel)
                .with_extension("exe")
                .to_string_lossy()
                .replace('/', "\\");

            append(
                &mut tasks_vs["tasks"],
                json!({
                    "taskLabel": format!("Build {name}"),
                    "appliesTo": "imp",
                    "type": "launch",
                    "command": "./imp",
                    "args": ["build", "--target", &target.name, "--mode", mode.to_string()],
                    "contextType": "build",
                    "output": format!("${{workspaceRoot}}\\{out_win}")
                }),
            );

            append(
                &mut launch_vs["configurations"],
                json!({
                    "type": "default",
                    "project": out_win,
                    "projectTarget": format!("View.Build ({}_{mode})", target.name),
                    "name": name,
                    "args": [],
                    "currentDir": "${workspaceRoot}"
                }),
            );

            append(
                &mut tasks_vsc["tasks"],
                json!({
                    "label": format!("Build {name}"),
                    "type": "shell",
                    "command": format!("./imp build --target {} --mode {mode}", target.name),
                    "group": "build",
                    "problemMatcher": []
                }),
            );

            let debugger = if cfg!(windows) {
                "cppvsdbg"
            } else {
                "lldb-dap"
            };
            append(
                &mut launch_vsc["configurations"],
                json!({
                    "name": name,
                    "type": debugger,
                    "request": "launch",
                    "program": format!("${{workspaceFolder}}/{}", out_rel),
                    "args": [],
                    "stopAtEntry": false,
                    "cwd": "${workspaceFolder}",
                    "environment": [],
                    "externalConsole": false,
                    "preLaunchTask": format!("Build {name}")
                }),
            );
        }
    }

    std::fs::write(
        vs_dir.join("launch.vs.json"),
        serde_json::to_string_pretty(&launch_vs)?,
    )?;
    std::fs::write(
        vs_dir.join("tasks.vs.json"),
        serde_json::to_string_pretty(&tasks_vs)?,
    )?;
    std::fs::write(
        vscode_dir.join("launch.json"),
        serde_json::to_string_pretty(&launch_vsc)?,
    )?;
    std::fs::write(
        vscode_dir.join("tasks.json"),
        serde_json::to_string_pretty(&tasks_vsc)?,
    )?;

    p.done("generated .vs/ and .vscode/");
    Ok(())
}

fn append(array: &mut Value, item: Value) {
    if let Value::Array(ref mut arr) = array {
        arr.push(item);
    }
}

File: rules/workflows/vs.js
// Ported from the legacy Rust command src/commands/vs.rs.
//
// Emits Visual Studio and VS Code build/launch configuration for every Odin
// package in the current selection, in Debug and Release variants:
//   .vs/launch.vs.json, .vs/tasks.vs.json, .vscode/launch.json, .vscode/tasks.json
//
// Its own "vs" goal, not "build" — emitting IDE config isn't building
// anything. Invoke as `imp vs //...` (or a narrower selector) — every Odin
// package exposing [VSCODE] in the selection is aggregated into one set of
// files. Graph-native: each package attaches its own [VSCODE] facet at
// declaration time (see //rules/odin/vscode), so this module has no central
// package registry of its own — it only aggregates whatever roots the
// engine resolves for the "vs" goal's selection.
//
// Output paths are computed with the same pure helper (odin_output_path,
// rules/odin/index.js) a real build uses internally — imported, not
// duplicated — so they can't drift from what a build actually produces.
// [VSCODE] roots only run a static source scan; no real build is ever
// triggered by this goal.

import { goal, run, output, output_path, platformInfo, group } from "imp:core";
import { odin_output_path, graphDefaultOutputPath } from "//rules/odin";

export const VSCODE = goal("vs", undefined, { graph: graphVsGoal });

const MODES = ["Debug", "Release"];

function targetName(address) {
	const colon = address.lastIndexOf(":");
	return colon >= 0 ? address.slice(colon + 1) : address;
}

function windowsOutput(outRel) {
	return `${outRel}.exe`.replace(/\//g, "\\");
}

function pretty(value) {
	return `${JSON.stringify(value, null, 2)}\n`;
}

// Write a JSON file as a cacheable run(). Content is passed as a positional
// argument so no shell interpolation touches it.
function writeJsonFile(path, value) {
	return run({
		argv: [
			"sh",
			"-c",
			'printf %s "$2" > "$1"',
			"vs-write",
			output_path(path),
			pretty(value),
		],
		outputs: [output(path)],
		materialize: true,
		display: `write ${path}`,
	});
}

/**
 * Aggregate every selected package's [VSCODE] facet into one set of
 * VS/VS Code IDE config files. Called by the engine with the resolved
 * roots for the "vs" goal's selection — no package registry involved.
 *
 * @param {Array<{address: string, result: {hasMainEntrypoint: boolean, packagePath: string}}>} roots
 */
export async function graphVsGoal(roots) {
	const isWindows = platformInfo().os === "windows";
	const debuggerType = isWindows ? "cppvsdbg" : "lldb-dap";

	const launchVs = { version: "0.2.1", defaults: {}, configurations: [] };
	const tasksVs = { version: "0.2.1", tasks: [] };
	const launchVsc = { version: "0.2.0", configurations: [] };
	const tasksVsc = { version: "2.0.0", tasks: [] };

	for (const { address, result: analysis } of roots) {
		const name = targetName(address);
		const out = graphDefaultOutputPath(address);
		const outRel = odin_output_path(out, analysis);
		const outWin = windowsOutput(outRel);
		const buildSelector = `${address}#build`;

		for (const mode of MODES) {
			const label = `${name} (${mode})`;
			const buildTask = `Build ${label}`;

			tasksVs.tasks.push({
				taskLabel: buildTask,
				appliesTo: "imp",
				type: "launch",
				command: "./imp",
				args: ["build", buildSelector],
				contextType: "build",
				output: `\${workspaceRoot}\\${outWin}`,
			});

			launchVs.configurations.push({
				type: "default",
				project: outWin,
				projectTarget: `View.Build (${name}_${mode})`,
				name: label,
				args: [],
				currentDir: "${workspaceRoot}",
			});

			tasksVsc.tasks.push({
				label: buildTask,
				type: "shell",
				command: `./imp build ${buildSelector}`,
				group: "build",
				problemMatcher: [],
			});

			launchVsc.configurations.push({
				name: label,
				type: debuggerType,
				request: "launch",
				program: `\${workspaceFolder}/${outRel}`,
				args: [],
				stopAtEntry: false,
				cwd: "${workspaceFolder}",
				environment: [],
				externalConsole: false,
				preLaunchTask: buildTask,
			});
		}
	}

	const files = {
		".vs/launch.vs.json": launchVs,
		".vs/tasks.vs.json": tasksVs,
		".vscode/launch.json": launchVsc,
		".vscode/tasks.json": tasksVsc,
	};
	await group(
		Object.entries(files).map(([path, value]) => writeJsonFile(path, value)),
	);

	return { generated: Object.keys(files) };
}
